### PR TITLE
Replaced ReadOnlySequence unsafe casting to safe casting in Seek method

### PR DESCRIPTION
**Pull request by AlexRadch**
Replaced ReadOnlySequence unsafe casting to safe casting in Seek method. Part of #28920

### CODE REVIEW

File: src/System.Memory/src/System/Buffers/ReadOnlySequence.cs [hunk 1]
```diff
@@ -315,15 +315,7 @@ public override string ToString()
                 if (Length < int.MaxValue)
                 {
 #if !FEATURE_PORTABLE_SPAN
-                    return string.Create((int)Length, charSequence, (span, sequence) =>
```
Comment (pakrym) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence.cs]: Nice!

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs [hunk 2]
```diff
@@ -195,7 +191,7 @@ private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> curr
             }
 
             // Hit the end of the segments but didn't reach the count
-            if (currentSegment == null || (currentSegment == endObject && endPosition < offset))
+            if (currentSegment == null || endIndex < offset)
```
Comment (pakrym) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: Why?

Comment (AlexRadch) [reply src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: While at above `while (currentSegment != null && currentSegment != endObject)` ensure that `currentSegment == null || currentSegment == endObject`.

So here `(currentSegment == null || (currentSegment == endObject && endPosition < offset))`
if `currentSegment == null` is false means that `currentSegment == endObject` so we can remove `currentSegment == endObject`.

Comment (AlexRadch) [reply src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: It can be rewrote as `currentSegment != endObject || endPosition < offset` also.

**Comment by ahsonkhan** [comment]
@dotnet-bot test OSX x64 Debug Build
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x86 Release Build

**Comment by pakrym** [comment]
How are Slice perf numbers after this?

**Comment by ahsonkhan** [comment]
> How are Slice perf numbers after this?

They are slightly better since we reduced removed some checks.

**Comment by pakrym** [comment]
Nice, LGTM.

**Comment by ahsonkhan** [comment]
This change was superseded by https://github.com/dotnet/corefx/pull/29135.

File: src/System.Memory/src/System/Buffers/ReadOnlySequence.cs [hunk 3]
```diff
@@ -315,15 +315,7 @@ public override string ToString()
...
-                    {
-                        foreach (ReadOnlyMemory<char> readOnlyMemory in sequence)
-                        {
-                            ReadOnlySpan<char> sourceSpan = readOnlyMemory.Span;
-                            sourceSpan.CopyTo(span);
-                            span = span.Slice(sourceSpan.Length);
-                        }
-                    });
+                    return string.Create((int)Length, charSequence, (span, sequence) => sequence.CopyTo(span));
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence.cs]: Please open a new PR with just this change since the fix to Seek is already in. Thanks.

**Comment by pakrym** [comment]
Closing as superseded by #29135. @AlexRadch can you please open a PR with ToString change?

**Pull request closed without merging**